suzieq/engines/pandas/interfaces.py

```diff
@@ -100,55 +100,85 @@ def _check_vlan_match(self, vlan_list: List[str],
         '''
 
-        # If there are any VLANs with ! mixed without !, remove the
-        # ! ones. vlanList is an OR list, not an AND list, except
-        # if all are !
-        if any(x.startswith('!') for x in vlan_list):
-            cond = 'and'
-        else:
-            cond = 'or'
-
-        resdf_list = []
-        if (any(x.startswith('>') for x in vlan_list) and
-                any(x.startswith('<') for x in vlan_list)):
-            # if the user specifies both >/>= and </<=, then we
-            # treat this as an and
-            cond = 'and'
-
         opdict = {'<': operator.lt, '>': operator.gt,
                   '<=': operator.le, '>=': operator.ge,
                   '!': operator.ne, '==': operator.eq}
 
-        for vlan in vlan_list:
-            if vlan.startswith(('<=', '>=')):
-                op = vlan[0:2]
-                vlan = vlan[2:].strip()
-            elif vlan.startswith(('<', '>', '!')):
-                op = vlan[0]
-                vlan = vlan[1:].strip()
-            else:
-                op = '=='
-
-            vlan = int(vlan)
-            if op == "!":
-                tmpdf = df[df.apply(
-                    lambda row, vlan: all(opdict[op](v, vlan)
-                                          for v in row.vlanList) and
-                    opdict[op](row.vlan, vlan), args=(vlan,), axis=1)]
-            else:
-                tmpdf = df[df.apply(
-                    lambda x, vlan: any(opdict[op](y, vlan)
-                                        for y in x.vlanList) or
-                    opdict[op](x.vlan, vlan), args=(vlan,), axis=1)]
-            if cond == "or":
-                resdf_list.append(tmpdf.reset_index(drop=True))
-            else:
-                df = tmpdf
-
-        if resdf_list:
-            df = pd.concat(resdf_list)
+        def _check_cond(vlist: pd.Series, op: Callable, *args):
+            fn = opdict[op]
+            return vlist.apply(lambda ls: any(fn(vlan, *args) for vlan in ls))
+
+        def _interval(ival, start_val, start_op, end_val, end_op):
+            start_op = opdict[start_op]
+            end_op = opdict[end_op]
+            return start_op(ival, start_val) and end_op(ival, end_val)
+
+        def extract_op(expression: Union[str, int]) -> Tuple[str, int]:
+            op = '=='
+            val = expression
+            if isinstance(expression, str):
+                if expression.startswith(('<=', '>=')):
+                    val = expression[2:]
+                    op = expression[:2]
+                elif expression.startswith(('<', '>')):
+                    val = expression[1:]
+                    op = expression[:1]
+                elif expression.startswith('!'):
+                    val = expression[1:]
+                    op = expression[:1]
+            return op, val
+
+        if not vlan_filters:
+            return df
 
-        return df.query('vlanList.str.len() > 0 or vlan != 0')
+        opdict.update({'><': _interval})
```

Comment:
why you update opdict instead of directly add this option at line 103?

Reply:
I know it's ugly but inside the _interval() function we use `opdict`, so when I define `opdict`,  `_interval()` has not been already defined